### ISSUE TITLE
Shadow Metadata is corrupted when Master crashes

### ISSUE THREAD

**Issue by nickcoons** [lizardfs/lizardfs#617]
I have a failover set up with uCarp between a metadata master and shadow server.  When I initiate a shutdown of the master server, uCarp runs some scripts that reconfigures the shadow as the new master, then restarts the service.  This all appears to work properly.

However, when the master server crashes, uCarp does its thing, but the former-shadow server will not restart as a master.  It complains that the metadata is corrupted and that I need to run mfsmetarestore (which then indicates that it's not able to resolve the issue).  Bringing the original master back up, then running mfsmetarestore brings LFS back up.

But my question is, why would my shadow metadata become corrupted when the master crashes?  This seems to happen consistently.  Am I misunderstanding how this is supposed to work?  That is, is the shadow supposed to be able to take over for a crashed master?  Or is it only supposed to work if I schedule a shutdown of master?

I'm currently running 3.9.4, but am planning for an upgrade in the near future.  I'm curious if this is:

1) A misunderstanding on my part of how it should work.
2) Something that's broken in 3.9.4 but has been fixed in a newer version.
3) Potentially something with my specific configuration (I'm trying to rule out #1 and #2 before spinning my wheels on my specific setup).

**Comment by 4Dolio** [comment]
You need to look into the personality=hamanaged mode. it allows a shadow to switch into master mode on the fly instantly with no reload.  You change personalities with a lfsadmin? command which can communicate directly with any master service to poll many cluster statuses and change modes from shadow to master... masters must stop start into new initial personality of shadow and can not do that live like a promotion can

**Comment by 4Dolio** [comment]
See https://github.com/lizardfs/lizardfs/issues/598#issuecomment-334202503

**Comment by nickcoons** [comment]
Here's how I have it set up now.

I have two configs for my metadata servers, `/etc/mfs/mfsmaster.master.cfg` and `/etc/mfs/mfsmaster.shadow.cfg`.  These files are identical, except the second one has `PERSONALITY = shadow` in it.  I have a script that puts a server into master mode, and another that puts it into shadow mode.  So for the server in master mode, I create a symlink called `/etc/mfs/mfsmaster.cfg` that points to the master config, and in shadow mode the symlink points to the shadow config.  After creating the appropriate symlink to put the server in the desired mode, it restarts the master service.

From what you're saying, it sounds like I can change the personality from master to shadow (or back) without restarting the service.  That sounds pretty cool, and it would simplify my configuration.   But do you think that it relates to the issue that I'm having?  Is my shadow server's metadata becoming corrupted because I'm changing the personality via the config file and then issuing `/etc/init.d/mfsmaster restart`?  Though it's not as simplistic as the method you've described, it doesn't seem that that would be my issue, does it?

**Comment by psarna** [comment]
3.9.4 had its issues, so please upgrade before taking any other steps. Your problem is most likely solved already in 3.11.3.

**Comment by nickcoons** [comment]
Do you think the issue is already solved in 3.10.4 as well?  I'm looking to upgrade my OS to Debian Stretch, which only has packages for 3.10.4 (I prefer to stay with the distribution packages for simpler management).

**Comment by psarna** [comment]
No idea. It makes sense to upgrade anyway, 3.10.4 contains a lot of fixes.

**Comment by Blackpaw** [comment]
>From what you're saying, it sounds like I can change the personality from master to shadow (or back) without restarting the service. 

I'm pretty sure demoting a master to shadow requires a restart of the service. However a shadow can be promoted to master by a simple reload.

**Comment by Zorlin** [comment]
@Blackpaw Yes. Demotion from master -> shadow via reload is supported, promotion via reload is NOT supported and likely never will be... a full restart is required in that case.

**Comment by Blackpaw** [comment]
>Demotion from master -> shadow via reload is supported, promotion via reload is NOT supported and likely never will be... a full restart is required in that case.

Isn't that the opposite of what I said?

**Comment by Zorlin** [comment]
@Blackpaw My mistake, sorry! Evidently I need more coffee... fixed.

**Comment by 4Dolio** [comment]
No. No no..

There is a third personality type. Of hamanaged, and that is the only personality you can use to switch personalities on the fly with.

If you swap configs between master/shadow then  all bets are off. You need to do a full cluster stop/start to change masters. Static personalities are not intended to be live transitioned.  Please refer to the other threads I mentioned, those all assume you use the 3rd peraonality=hamanaged

**Comment by nickcoons** [comment]
@4Dolio, I'm a little confused, because it sounds like there are conflicting claims in this thread about that.

So if I set personality=hamanaged, what is the personality of a metadata server when it starts, master or shadow?  Can a metadata server be set this way, and then promoted or demoted without a reload of the service?

**Comment by 4Dolio** [comment]
Yes, we are inducing a fair amount of confusion in this thread.  I believe I am correct in that of the three personalities, the master/shadow pair are for a non-HA cluster where you manually cold stop/start to change your master (Not that you can not do this quickly, but it is still full stop/start).  Only the hamanaged personality is intended to do live HA personality promotions (demotions still stop/start, because, well, that's silly to want to live demote a master, see https://github.com/lizardfs/lizardfs/issues/613#issuecomment-340287341 and my replies)

To start a master node with personality=hamanaged you have to tell it the personality you want when you start it, see:
https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in#L216

It would look something like this if you just did it by hand with no $vars:
`mfsmaster -c "/etc/mfs/mfsmaster.cfg" -o ha-cluster-managed -o initial-personality="master" "start"`

And here is the promotion function:
https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in#L325
Here is the call that attempts to execute the promotion
https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in#L482

And demotions:
https://github.com/4Dolio/lizardfs/blob/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster/metadataserver.in#L438

Hope this helps give you enough information to look into for your HA solution.  Happy to help once you get up to speed with the hamanaged personality mechanics..

**Comment by Blackpaw** [comment]
>There is a third personality type. Of hamanaged, and that is the only personality you can use to switch personalities on the fly with.

I don't think so, its not a real personality type, just a useful config option for use with lizardadm. I'm pretty sure masters can never be safely demoted to shadows.

**Comment by 4Dolio** [comment]
Shrugs. Idk, only use hamanaged.

I concur, demotion can never be done with live reload, only restart or stop/start. (I hope I never stated the contrary).  Only a promotion can happen live with a reload or promote call.

**Comment by guestisp** [comment]
Let's assume shadow1 and shadow2 are connected to a master
Master fails, shadow1 will be automatically promoted to master, what happens to shadow2? Will be reconnected to the new master and still syncing?

**Comment by Blackpaw** [comment]
If you are using a floating ip that is now assigned to the new master, then yes shadow2 will reconnect, along with all the chunkserevrs, mounts etc.
